src/graph/graph.c

```diff
@@ -680,10 +674,288 @@ void Graph_DeleteNode(Graph *g, Node *n) {
         GxB_Matrix_Delete(M, ENTITY_GET_ID(n), ENTITY_GET_ID(n));
     }
 
-    FreeEntity(n->entity);
     DataBlock_DeleteItem(g->nodes, ENTITY_GET_ID(n));
 }
 
+void _BulkDeleteNodes(Graph *g, Node *nodes, size_t node_count, size_t *edge_deleted) {
+    assert(g && g->_writelocked && nodes && node_count > 0);
+
+    /* Create a matrix M where M[j,i] = 1 where:
+     * Node i in is connected to node j. */
+    
+    GrB_Matrix A;                       // A = R(M) masked relation matrix.
+    GrB_Matrix Mask;                    // Mask noteing all implicitly deleted edges.
+    GrB_Matrix Nodes;                   // Mask noteing each node marked for deletion.
+    GrB_Matrix adj;                     // Adjacency matrix.
+    GrB_Matrix tadj;                    // Transposed adjacency matrix.
+    GrB_Descriptor desc;                // GraphBLAS descriptor.
+    GxB_SelectOp selectop;              // GraphBLAS select operator, used to free edges.
+    GxB_MatrixTupleIter *adj_iter;      // iterator over the adjacency matrix.
+    GxB_MatrixTupleIter *tadj_iter;     // iterator over the transposed adjacency matrix.
+
+    GrB_Descriptor_new(&desc);
+    adj = Graph_GetAdjacencyMatrix(g);
+    tadj = _Graph_Get_Transposed_AdjacencyMatrix(g);
+    GxB_MatrixTupleIter_new(&adj_iter, adj);
+    GxB_MatrixTupleIter_new(&tadj_iter, tadj);
+    GxB_SelectOp_new(&selectop, _select_op_free_edge, GrB_UINT64);
+    GrB_Matrix_new(&A, GrB_UINT64, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));
+    GrB_Matrix_new(&Mask, GrB_BOOL, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));    
+    GrB_Matrix_new(&Nodes, GrB_BOOL, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));
+
+    // Populate mask with implicit edges, take note of deleted nodes.
+    for(uint i = 0; i < node_count; i++) {
```

Comment:
It's a quibble, but I don't love how inconsistent our int types are. `node_count` is a size_t, though it should actually be a uint like `i`.

Reply:
👍